include/deal.II/matrix_free/vector_access_internal.h

```diff
@@ -0,0 +1,683 @@
+                           const types::global_dof_index          entry,
+                           const typename VectorType::value_type &val)
+  {
+    vec(entry) += val;
+  }
+
+
+
+  template <typename VectorType,
+            typename std::enable_if<has_set_local_element<VectorType>::value,
+                                    VectorType>::type * = nullptr>
+  inline void
+  vector_access_set(VectorType &                           vec,
+                    const unsigned int                     entry,
+                    const typename VectorType::value_type &val)
+  {
+    vec.set_local_element(entry, val);
+  }
+
+
+
+  template <typename VectorType,
+            typename std::enable_if<!has_set_local_element<VectorType>::value,
+                                    VectorType>::type * = nullptr>
+  inline void
+  vector_access_set(VectorType &                           vec,
+                    const unsigned int                     entry,
+                    const typename VectorType::value_type &val)
+  {
+    vector_access(vec, entry) = val;
+  }
+
+
+
+  // this is to make sure that the parallel partitioning in VectorType
+  // is really the same as stored in MatrixFree.
+  // version below is when has_partitioners_are_compatible == false
+  // FIXME: this is incorrect for PETSc/Trilinos MPI vectors
+  template <
+    typename VectorType,
+    typename std::enable_if<!has_partitioners_are_compatible<VectorType>::value,
+                            VectorType>::type * = nullptr>
+  inline void
+  check_vector_compatibility(
+    const VectorType &                            vec,
+    const internal::MatrixFreeFunctions::DoFInfo &dof_info)
+  {
+    (void)vec;
+    (void)dof_info;
+
+    AssertDimension(vec.size(), dof_info.vector_partitioner->size());
+  }
+
+
+
+  // same as above for has_partitioners_are_compatible == true
+  template <
+    typename VectorType,
+    typename std::enable_if<has_partitioners_are_compatible<VectorType>::value,
+                            VectorType>::type * = nullptr>
+  inline void
+  check_vector_compatibility(
+    const VectorType &                            vec,
+    const internal::MatrixFreeFunctions::DoFInfo &dof_info)
+  {
+    (void)vec;
+    (void)dof_info;
+    Assert(vec.partitioners_are_compatible(*dof_info.vector_partitioner),
+           ExcMessage(
+             "The parallel layout of the given vector is not "
+             "compatible with the parallel partitioning in MatrixFree. "
+             "Use MatrixFree::initialize_dof_vector to get a "
+             "compatible vector."));
+  }
+
+
+
+  // Below, three classes (VectorReader, VectorSetter,
+  // VectorDistributorLocalToGlobal) implement the same interface and can be
+  // used to to read from vector, set elements of a vector and add to elements
+  // of the vector.
+
+  // 1. A class to read data from vector
+  template <typename Number, typename VectorizedArrayType>
+  struct VectorReader
+  {
```

Comment:
I like these new classes! I will have a second look a these later, since I think I might be able to use it and be able to replace a class (700 locs) from our application code (FYI @kkormann).

Reply:
They are not new, though, I just moved them from the `fe_evaluation.h` to this one.